File: frontend-map/lib/geo.ts
// Mirrors backend-watchlist/app/services/geo.py exactly -- both need the
// same haversine/bearing math because bearing/speed only arrive from the
// backend on the scripted vehicle-trace-demo path (GET /vehicle-traces);
// a normal plate search (GET /detections, camera resolved client-side from
// the frontend's own registry) has real coordinates too but no server-side
// enrichment, so buildSightingRoute computes it here instead. Investigators
// searching a real plate get the same "inferred direction of travel" and
// average speed the demo path shows, not a lesser version of the feature.
const EARTH_RADIUS_KM = 6371.0088;

function toRad(deg: number): number {
  return (deg * Math.PI) / 180;
}

export function haversineKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const phi1 = toRad(lat1);
  const phi2 = toRad(lat2);
  const dPhi = toRad(lat2 - lat1);
  const dLambda = toRad(lon2 - lon1);
  const a =
    Math.sin(dPhi / 2) ** 2 + Math.cos(phi1) * Math.cos(phi2) * Math.sin(dLambda / 2) ** 2;
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(a));
}

export function initialBearingDeg(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const phi1 = toRad(lat1);
  const phi2 = toRad(lat2);
  const dLambda = toRad(lon2 - lon1);
  const x = Math.sin(dLambda) * Math.cos(phi2);
  const y = Math.cos(phi1) * Math.sin(phi2) - Math.sin(phi1) * Math.cos(phi2) * Math.cos(dLambda);
  const theta = Math.atan2(x, y);
  return ((theta * 180) / Math.PI + 360) % 360;
}

/** bearing/speed for the leg from `prev` to `curr`. null/null when the two
 * timestamps don't actually advance (dividing by a non-positive duration
 * would produce a meaningless or negative speed). */
export function legBearingAndSpeed(
  prev: { lat: number; lon: number; detectedAt: string },
  curr: { lat: number; lon: number; detectedAt: string }
): { bearingDeg: number; speedKmh: number | null; gapHours: number } {
  const bearingDeg = initialBearingDeg(prev.lat, prev.lon, curr.lat, curr.lon);
  const dtHours =
    (new Date(curr.detectedAt).getTime() - new Date(prev.detectedAt).getTime()) / 1000 / 3600;
  if (dtHours <= 0) return { bearingDeg, speedKmh: null, gapHours: dtHours };

  const distanceKm = haversineKm(prev.lat, prev.lon, curr.lat, curr.lon);
  return { bearingDeg, speedKmh: Math.round((distanceKm / dtHours) * 10) / 10, gapHours: dtHours };
}

// Mirrors backend-watchlist/app/services/geo.py's classify_leg_anomaly and
// its two constants exactly -- see that module's docstring for why these
// are heuristics ("review this leg"), not findings.
export const MAX_PLAUSIBLE_SPEED_KMH = 160.0;
export const EXTENDED_GAP_HOURS = 12.0;
export const IMPROBABLE_SPEED = 'improbable_speed';
export const EXTENDED_GAP = 'extended_gap';

export function classifyLegAnomaly(speedKmh: number | null, gapHours: number | null): string | null {
  if (speedKmh != null && speedKmh > MAX_PLAUSIBLE_SPEED_KMH) return IMPROBABLE_SPEED;
  if (gapHours != null && gapHours > EXTENDED_GAP_HOURS) return EXTENDED_GAP;
  return null;
}

File: backend-watchlist/app/services/geo.py
"""Pure geometry helpers for the vehicle-trace trajectory view -- no DB
access, so these are trivial to unit-test in isolation from the rest of
detections_service.

Direction/speed here are inferred from the order and spacing of fixed ANPR
camera reads, the same way large-scale ALPR platforms (e.g. Genetec AutoVu's
ML Core) report "direction of travel" and average speed between fixed
checkpoints -- not continuous GPS, just consecutive known points and a time
delta. The frontend labels this "inferred", never "GPS tracking", to keep
that distinction honest to an investigator reading it.
"""
import math

EARTH_RADIUS_KM = 6371.0088

# Anomaly heuristics -- deliberately simple, deliberately named as heuristics
# rather than "detected fraud": a real investigator treats these as "review
# this leg", not proof of anything. Two signals a plain route-on-a-map never
# surfaces, both explicitly called out in the brief this feature is built
# against ("flagging... suspicious route anomalies"):
#   - IMPROBABLE_SPEED: the two reads imply a speed no real vehicle covers
#     that ground distance in that time on Indian roads -- almost always a
#     wrong-plate OCR match (two different vehicles' reads merged into one
#     plate's history) rather than an actual 160+ km/h drive between fixed
#     checkpoints.
#   - EXTENDED_GAP: the plate vanishes for a long stretch between two
#     otherwise-active checkpoints -- not proof of anything by itself (it
#     could've simply left the ANPR-covered road network), but exactly the
#     kind of gap worth a human glancing at when reviewing a plate's history.
MAX_PLAUSIBLE_SPEED_KMH = 160.0
EXTENDED_GAP_HOURS = 12.0

IMPROBABLE_SPEED = "improbable_speed"
EXTENDED_GAP = "extended_gap"


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Great-circle distance between two lat/long points, in kilometers."""
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlambda = math.radians(lon2 - lon1)
    a = math.sin(dphi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(dlambda / 2) ** 2
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(a))


def initial_bearing_deg(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Compass bearing (0-360, 0 = north) from point 1 to point 2."""
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    dlambda = math.radians(lon2 - lon1)
    x = math.sin(dlambda) * math.cos(phi2)
    y = math.cos(phi1) * math.sin(phi2) - math.sin(phi1) * math.cos(phi2) * math.cos(dlambda)
    theta = math.atan2(x, y)
    return (math.degrees(theta) + 360) % 360


def leg_bearing_and_speed(prev: dict, curr: dict) -> tuple[float | None, float | None]:
    """bearing_deg/speed_kmh for the leg from `prev` to `curr`, each a dict
    with latitude/longitude/detected_at. None/None when either point lacks
    coordinates, or the two timestamps don't actually advance (a replayed/
    out-of-order pair -- dividing by a non-positive duration would produce a
    meaningless or negative speed)."""
    if None in (prev.get("latitude"), prev.get("longitude"), curr.get("latitude"), curr.get("longitude")):
        return None, None

    bearing = initial_bearing_deg(prev["latitude"], prev["longitude"], curr["latitude"], curr["longitude"])
    dt_hours = (curr["detected_at"] - prev["detected_at"]).total_seconds() / 3600
    if dt_hours <= 0:
        return bearing, None

    distance_km = haversine_km(prev["latitude"], prev["longitude"], curr["latitude"], curr["longitude"])
    return bearing, round(distance_km / dt_hours, 1)


def classify_leg_anomaly(speed_kmh: float | None, gap_hours: float | None) -> str | None:
    """Which heuristic (if any) this leg trips, checked in this order:
    an improbable speed is the stronger, more specific signal (implies a
    likely OCR mismatch) and takes priority over a merely-long gap when a
    leg somehow trips both."""
    if speed_kmh is not None and speed_kmh > MAX_PLAUSIBLE_SPEED_KMH:
        return IMPROBABLE_SPEED
    if gap_hours is not None and gap_hours > EXTENDED_GAP_HOURS:
        return EXTENDED_GAP
    return None
